### PR TITLE
Intercept poststop

### PR DESCRIPTION
**Pull request by johanandren**
Refs #26285 

Competes with #26292

### CODE REVIEW

**Comment by johanandren** [comment]
Also covers #26293

File: akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala [hunk 1]
```diff
@@ -261,16 +260,26 @@ object Behavior {
    * INTERNAL API: When the cell is stopping this behavior is used, so
    * that PostStop can be sent to previous behavior from `finishTerminate`.
    */
-  private[akka] class StoppedBehavior[T](val postStop: OptionVal[Behavior[T]]) extends Behavior[T] {
-    validatePostStop(postStop)
+  private[akka] sealed class StoppedBehavior[T](val postStop: OptionVal[TypedActorContext[T] ⇒ Unit]) extends ExtensibleBehavior[T] {
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala]: This is half of the essence of this PR - make StoppedBehavior an extensible behavior, and implement the callback as a normal signal (which can come through the interceptor)

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala [hunk 2]
```diff
@@ -84,7 +83,11 @@ private[akka] final class InterceptorImpl[O, I](val interceptor: BehaviorInterce
 
   private def deduplicate(interceptedResult: Behavior[I], ctx: TypedActorContext[O]): Behavior[O] = {
     val started = Behavior.start(interceptedResult, ctx.asInstanceOf[TypedActorContext[I]])
-    if (started == UnhandledBehavior || started == SameBehavior || !Behavior.isAlive(started)) {
+    if (Behavior.isStopped(started)) {
+      // keep the interceptor around stopped behaviors so that it can intercept
+      // the post-stop signal
+      new InterceptorImpl(interceptor, started)
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala]: Make sure we keep the interceptor when the behavior is stopped (but not for `Failed`)

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala [hunk 3]
```diff
@@ -86,22 +85,25 @@ import akka.util.OptionVal
...
+              throw new IllegalStateException("Stopping, should never receieve a message")
+            override def receiveSignal(ctx: TypedActorContext[T], msg: Signal): Behavior[T] = {
+              // FIXME the old logic checked if there is a stopped callback here, and only if there
+              // wasn't it would PostStop signal the previous behavior, now the stop behavior
+              // may be wrapped in any number of interceptors though, so doing that is tricky
+              // therefore currently both the previous behavior and the returned get the signal
+              // maybe that is ok, or maybe it is problematic, not sure
+              Behavior.interpretSignal(lastBehavior, ctx, PostStop)
+              Behavior.interpretSignal(s, ctx, PostStop)
+            }
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: Since the behavior can be wrapped in any number of interceptors, it is hard to pull the OptionVal callback out of there, so instead I made it invoke both that and the callback.

This means such a behavior:
```
Behavior.receiveMessage { 
  case "stop" =>
    Behavior.stop(() => bö)
} receiveSignal {
  case (_, PostStop) =>
    bo
}
```

Will execute both `bo` and `bö`. This didn't break any tests but maybe there are scenarios where there is a problem.

Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: That doesn't look too bad to me, given that it is now a callback.

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala [hunk 4]
```diff
@@ -78,7 +78,7 @@ object Behaviors {
    * given `postStop` behavior. All other messages and signals will effectively be
    * ignored.
    */
-  def stopped[T](postStop: Behavior[T]): Behavior[T] = Behavior.stopped(postStop)
+  def stopped[T](postStop: Effect): Behavior[T] = Behavior.stopped(postStop.apply)
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala]: update docs

File: akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Behaviors.scala [hunk 5]
```diff
@@ -66,7 +66,7 @@ object Behaviors {
    * given `postStop` behavior. All other messages and signals will effectively be
    * ignored.
    */
-  def stopped[T](postStop: Behavior[T]): Behavior[T] = Behavior.stopped(postStop)
+  def stopped[T](postStop: () ⇒ Unit): Behavior[T] = Behavior.stopped(postStop)
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Behaviors.scala]: update docs

**Review by patriknw — COMMENTED** [review]
this looks promising

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala (hunk 3 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: Yeah, I have thought about it since pushing and I can't think of any scenario where it is very surprising. Let's go with this!

**Comment by johanandren** [comment]
PLS BUILD

**Comment by johanandren** [comment]
Looks like it in fact changed ordering of something in typed persistence, can't seem to repeat locally though, so a race of some kind...

**Comment by johanandren** [comment]
I can in fact repeat locally, was testing typed persistence but issue is in typed sharding. 

Reminder to myself: The issue is something with the `PoisonPillInterceptor` for sharding vs this new `PostStop` interception logic.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala (hunk 3 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: I found a scenario where it is surprising now - with an interceptor on the previous behavior, the interceptor is kept to the stopping behavior and will get the PostStop on both. I haven't quite gotten my head around this issue - is it ok - should we make sure it doesn't happen etc.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala (hunk 2 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala]: Maybe we just shouldn't do this part - this is what causes the double-Signal in persistence.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/InterceptSpec.scala [hunk 6]
```diff
@@ -341,6 +341,37 @@ class InterceptSpec extends ScalaTestWithActorTestKit(
       probe.expectMessageType[B]
     }
 
+    "intercept PostStop" in {
+      @volatile var interceptorSawPostStop = false
+      @volatile var callbackSawPostStop = false
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/InterceptSpec.scala]: I'd probably use ordinary `TestProbe`s for these

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/GracefulStopSpec.scala [hunk 7]
```diff
@@ -54,13 +52,10 @@ final class GracefulStopSpec extends ScalaTestWithActorTestKit with WordSpecLike
...
-            Behaviors.receiveSignal {
-              case (_, PostStop) ⇒
-                // cleanup function body
-                probe.ref ! Done
-                Behaviors.same
-            }
+          Behaviors.stopped { () ⇒
+            // cleanup function body
+            probe.ref ! Done
+            Behaviors.same
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/GracefulStopSpec.scala]: isn't this `() => ()` now?

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/GracefulStopSpec.scala [hunk 8]
```diff
@@ -32,14 +32,12 @@ final class GracefulStopSpec extends ScalaTestWithActorTestKit with WordSpecLike
...
-            Behaviors.receiveSignal {
-              case (_, PostStop) ⇒
-                // cleanup function body
-                probe.ref ! "parent-done"
-                Behaviors.same
-            }
+          Behaviors.stopped { () ⇒
+            // cleanup function body
+            probe.ref ! "parent-done"
+            Behaviors.same
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/GracefulStopSpec.scala]: isn't this `() => ()` now?

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/InterceptSpec.scala [hunk 9]
```diff
@@ -341,6 +341,37 @@ class InterceptSpec extends ScalaTestWithActorTestKit(
...
+              interceptorSawPostStop = true
+          }
+          target(ctx, signal)
+        }
+      }
+
+      val ref = spawn(Behaviors.intercept(postStopInterceptor)(Behaviors.receiveMessage[String] { _ ⇒
+        Behaviors.stopped { () ⇒
+          callbackSawPostStop = true
+          Behaviors.same
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/InterceptSpec.scala]: isn't this `() => ()` now?

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/PoisonPill.scala [hunk 10]
```diff
@@ -12,6 +12,9 @@ import akka.annotation.InternalApi
 
 /**
  * INTERNAL API
+ *
+ * Note that this is not a universal poison pill like the untyped actor one but requires special handling
+ * on the receiving side where it is used (for example with the interceptor below.
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/PoisonPill.scala]: 👍

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala [hunk 11]
```diff
@@ -200,11 +190,9 @@ import akka.util.OptionVal
       case null                   ⇒ // skip PostStop
       case _: DeferredBehavior[_] ⇒
       // Do not undefer a DeferredBehavior as that may cause creation side-effects, which we do not want on termination.
-      case s: StoppedBehavior[_] ⇒ s.postStop match {
-        case OptionVal.Some(postStop) ⇒ Behavior.interpretSignal(postStop, ctx, PostStop)
-        case OptionVal.None           ⇒ // no postStop behavior defined
-      }
-      case b ⇒ Behavior.interpretSignal(b, ctx, PostStop)
+      case b ⇒
+        if (_ctx eq null) initializeContext()
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: why is this needed? should it always be done in `preStart` instead?

**Review by patriknw — COMMENTED** [review]
I haven't understood this yet

**Review by patriknw — COMMENTED** [review]
maybe cleanup this PR before I take another look, I'm not sure I see what questions/trouble is?

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/GracefulStopSpec.scala (hunk 7 above)
Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/GracefulStopSpec.scala]: Indeed. Leftover.

File: akka-stream-tests/src/test/scala/akka/stream/impl/TraversalBuilderSpec.scala [hunk 12]
```diff
@@ -463,7 +463,7 @@ class TraversalBuilderSpec extends AkkaSpec {
     //TODO: Dummy test cases just for smoke-testing. Should be removed.
 
     "foo" in {
-      implicit val mat = PhasedFusingActorMaterializer(ActorMaterializerSettings(system).withSyncProcessingLimit(5000))
+      implicit val mat = PhasedFusingActorMaterializer()
```
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-stream-tests/src/test/scala/akka/stream/impl/TraversalBuilderSpec.scala]: Huh, strange. Maybe I messed up with rebasing at some point.

**Comment by johanandren** [comment]
I'll rebase and cleanup.

In the meanwhile the issue I'm pondering is:

 1. To be able to intercept `PostStop` we keep the interception around a stopped behavior. 
 1. To be able to deal with `PostStop` both as a signal to previous behavior and to `Stopped` (to invoke the callback) we pass the signal to both those. 

This means that if the previous behavior was intercepted (like with persistence), that interceptor will get `PostStop` for both the previous behavior and the new `Stopped` behavior.

I'm trying to figure out if there is a way around it, but haven't gotten there yet.

**Comment by johanandren** [comment]
I think we can in fact not intercept around `StoppedBehavior` and just rely on the fact that previous behavior (which is intercepted) get the `PostStop`.

Which maybe makes a part of this PR non-necessary, because `StoppedBehavior` wouldn't have to be an `ExtensibleBehavior` if we don't need to be able to intercept around it. Hmmmm. I'll see if I can simplify based on this insight.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala (hunk 11 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: Yeah, makes sense to me, seems like we did some premature optimization on avoiding to initialize ctx in case of stopping immediately but that seems like a weird case to optimize to me.

**Comment by johanandren** [comment]
Tried removing that context initialization, but it is really intricate with how it is expected to happen, because of the guardian adapter subtype, I wonder if we can defer that for now.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StopSpec.scala [hunk 13]
```diff
@@ -21,22 +20,49 @@ class StopSpec extends ScalaTestWithActorTestKit with WordSpecLike {
 
   "Stopping an actor" should {
 
-    "execute the post stop" in {
+    "execute the post stop when stopping after setup" in {
       val sawSignal = Promise[Done]()
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StopSpec.scala]: why use different approaches in different tests? I'd prefer TestProbe

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StopSpec.scala [hunk 14]
```diff
@@ -21,22 +20,49 @@ class StopSpec extends ScalaTestWithActorTestKit with WordSpecLike {
...
+          Behaviors.same
+      })
+      ref ! "stop"
+      probe.expectMessage("signal")
+      probe.expectMessage("callback")
+    }
+
     // #25082
     "execute the post stop when wrapped" in {
       val sawSignal = Promise[Done]()
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StopSpec.scala]: ^

**Review by patriknw — APPROVED** [review]
LGTM, with testing nitpick

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StopSpec.scala (hunk 13 above)
Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StopSpec.scala]: No real motivation, I'll change.

**Comment by johanandren** [comment]
Ready for final review

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/TimerSpec.scala [hunk 15]
```diff
@@ -8,6 +8,8 @@ import java.util.concurrent.CountDownLatch
 import java.util.concurrent.TimeUnit
 import java.util.concurrent.atomic.AtomicInteger
 
+import akka.actor.DeadLetter
```
Comment (helena) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/TimerSpec.scala]: nit: keep akka imports together, I'd keep it under the languages.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/TimerSpec.scala [hunk 16]
```diff
@@ -74,10 +76,10 @@ class TimerSpec extends ScalaTestWithActorTestKit(
           throw e
       }
     } receiveSignal {
-      case (context, PreRestart) ⇒
+      case (_, PreRestart) ⇒
         monitor ! GotPreRestart(timer.isTimerActive("T"))
         Behaviors.same
-      case (context, PostStop) ⇒
```
Comment (helena) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/TimerSpec.scala]: fixing compiler errors? nice!

File: akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala [hunk 17]
```diff
@@ -261,22 +262,19 @@ object Behavior {
    * INTERNAL API: When the cell is stopping this behavior is used, so
    * that PostStop can be sent to previous behavior from `finishTerminate`.
    */
-  private[akka] class StoppedBehavior[T](val postStop: OptionVal[Behavior[T]]) extends Behavior[T] {
-    validatePostStop(postStop)
+  private[akka] sealed class StoppedBehavior[T](val postStop: OptionVal[TypedActorContext[T] ⇒ Unit]) extends Behavior[T] {
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala]: 👍

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/PoisonPill.scala [hunk 18]
```diff
@@ -12,6 +12,9 @@ import akka.annotation.InternalApi
 
 /**
  * INTERNAL API
+ *
+ * Note that this is not a universal poison pill like the untyped actor one but requires special handling
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/PoisonPill.scala]:
```suggestion
 * Note that this is a `Signal` poison pill, not a universal poison pill like the untyped actor one. This requires special handling.
```

?

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala [hunk 19]
```diff
@@ -86,22 +87,12 @@ import akka.util.OptionVal
...
-              // use the given postStop behavior, but canonicalize it
-              behavior = new Behavior.StoppedBehavior(OptionVal.Some(Behavior.canonicalize(postStop, behavior, ctx)))
-          }
-          context.stop(self)
         case f: FailedBehavior ⇒
           // For the parent untyped supervisor to pick up the exception
           throw new TypedActorFailedException(f.cause)
+        case stopped: StoppedBehavior[T] ⇒
+          behavior = new ComposedStoppingBehavior[T](behavior, stopped)
+          context.stop(self)
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: So much cleaner 👍

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala [hunk 20]
```diff
@@ -135,9 +126,10 @@ import akka.util.OptionVal
   }
 
   override def unhandled(msg: Any): Unit = msg match {
-    case Terminated(ref) ⇒ throw DeathPactException(ref)
-    case _: Signal       ⇒ // that's ok
-    case other           ⇒ super.unhandled(other)
+    case Terminated(ref) ⇒
+      throw DeathPactException(ref)
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: Same line like the others

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/PoisonPill.scala [hunk 21]
```diff
@@ -12,6 +12,9 @@ import akka.annotation.InternalApi
...
+ * on the receiving side where it is used (for example with the interceptor below.
  */
 @InternalApi private[akka] sealed abstract class PoisonPill extends Signal
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/PoisonPill.scala]: I'm sure not, for several reasons, but bringing up anyway?
```scala
@InternalApi private[typed] sealed abstract class PoisonPill extends Signal
```

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala [hunk 22]
```diff
@@ -260,3 +248,24 @@ private[typed] class GuardianActorAdapter[T](_initialBehavior: Behavior[T]) exte
...
 }
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[typed] final class ComposedStoppingBehavior[T](
+  lastBehavior: Behavior[T],
+  stopBehavior: StoppedBehavior[T]) extends ExtensibleBehavior[T] {
+  override def receive(ctx: TypedActorContext[T], msg: T): Behavior[T] =
+    throw new IllegalStateException("Stopping, should never receieve a message")
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: > should never receieve a message

This is private[typed] vs user facing. We have to satisfy the parent requirements, disallow shooting oneself in the foot, but reacting with throwing versus returning what is declared, some kind of `Behavior[T]` ... can we perhaps return a `FailedBehavior` or a new `UnsupportedBehavior[T]` or `UnexpectedBehavior[T]` ?

Comment (helena) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: I agree with _fail fast_ but...perhaps there's another way.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala [hunk 23]
```diff
@@ -260,3 +248,24 @@ private[typed] class GuardianActorAdapter[T](_initialBehavior: Behavior[T]) exte
...
+  override def receiveSignal(ctx: TypedActorContext[T], msg: Signal): Behavior[T] = {
+    if (msg != PostStop)
+      throw new IllegalArgumentException(s"The ComposedStoppingBehavior should never receive any other signal than PostStop but got $msg")
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: Same thought as above.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala (hunk 22 above)
Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: I think it would be a bug in Akka if it occured, and therefore the throwing

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/PoisonPill.scala (hunk 21 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/PoisonPill.scala]: Yeah, I'd probably have done narrower access control but Akka tradition [tm] is to keep them project private rather.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala (hunk 22 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: Yeah, it would be a bug in Akka, not anything the user could cause.

**Comment by johanandren** [comment]
Fail is jenkins server maintenance.

**Comment by raboof** [comment]
PLS BUILD

**Comment by johanandren** [comment]
Yay green!

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/TimerSpec.scala [hunk 24]
```diff
@@ -316,5 +318,23 @@ class TimerSpec extends ScalaTestWithActorTestKit("""
...
+          }
+        }
+      })
+      EventFilter.info("stopping").intercept {
+        ref ! "stop"
+      }
+      probe.expectTerminated(ref)
+      system.toUntyped.eventStream.subscribe(probe.ref.toUntyped, classOf[DeadLetter])
+      probe.expectNoMessage(1.second)
+    }
```
Comment (johanandren) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/TimerSpec.scala]: This was the reported issue, since PostStop never was intercepted the timer was never cancelled.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala [hunk 25]
```diff
@@ -287,3 +273,25 @@ private[typed] class GuardianActorAdapter[T](_initialBehavior: Behavior[T]) exte
...
+        s"The ComposedStoppingBehavior should never receive any other signal than PostStop but got $msg")
+    // first pass the signal to the previous behavior, so that it and potential interceptors
+    // will get the PostStop signal, unless it is deferred, we don't start a behavior while stopping
+    if (!lastBehavior.isInstanceOf[DeferredBehavior[_]])
+      Behavior.interpretSignal(lastBehavior, ctx, PostStop)
+    // and then to the potential stop hook, which can have a call back or not
+    stopBehavior.onPostStop(ctx)
+    Behavior.empty
+  }
+}
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: This is essentially the solution: we always pass the signal to the behavior before stopping, before we invoke the PostStop hook (which was before not a callback but a behavior).

This means the interceptors of the previous behavior get the `PostStop` signal, without having to wrap also `StoppedBehavior` with the interceptor.

Previous behavior was to only pass `PostStop` to the previous behavior if there was no callback in the `StoppedBehavior`

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala [hunk 26]
```diff
@@ -74,11 +74,11 @@ object Behaviors {
...
    *
-   * The PostStop signal that results from stopping this actor will be passed to the
-   * given `postStop` behavior. All other messages and signals will effectively be
-   * ignored.
+   * The `PostStop` signal that results from stopping this actor will first be passed to the
+   * current behavior and then the provided `postStop` callback will be invoked.
+   * All other messages and signals will effectively be ignored.
    */
-  def stopped[T](postStop: Behavior[T]): Behavior[T] = Behavior.stopped(postStop)
+  def stopped[T](postStop: Effect): Behavior[T] = Behavior.stopped(postStop.apply _)
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala]: This is an additional improvement, that is not 100% overlapping with the issue, but made the issue easier to solve - before the callback for  `stopped` was a behavior that would only ever get the `PostStop` signal, now it is a callback.

**Review by johanandren — COMMENTED** [review]
This is ready for final review!

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala (hunk 22 above)
Comment (helena) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: Sounds good

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala [hunk 27]
```diff
@@ -287,3 +273,25 @@ private[typed] class GuardianActorAdapter[T](_initialBehavior: Behavior[T]) exte
...
+ */
+@InternalApi private[typed] final class ComposedStoppingBehavior[T](lastBehavior: Behavior[T],
+                                                                    stopBehavior: StoppedBehavior[T])
+    extends ExtensibleBehavior[T] {
+  override def receive(ctx: TypedActorContext[T], msg: T): Behavior[T] =
+    throw new IllegalStateException("Stopping, should never receieve a message")
+  override def receiveSignal(ctx: TypedActorContext[T], msg: Signal): Behavior[T] = {
+    if (msg != PostStop)
+      throw new IllegalArgumentException(
+        s"The ComposedStoppingBehavior should never receive any other signal than PostStop but got $msg")
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]:
```suggestion
        s"The ComposedStoppingBehavior should only receive a PostStop signal, but received $msg")
```

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala [hunk 28]
```diff
@@ -287,3 +273,25 @@ private[typed] class GuardianActorAdapter[T](_initialBehavior: Behavior[T]) exte
...
+    // first pass the signal to the previous behavior, so that it and potential interceptors
+    // will get the PostStop signal, unless it is deferred, we don't start a behavior while stopping
+    if (!lastBehavior.isInstanceOf[DeferredBehavior[_]])
+      Behavior.interpretSignal(lastBehavior, ctx, PostStop)
```
Comment (helena) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala]: Without `isInstanceOf`?
```scala
lastBehavior match {
  case DeferredBehavior[_] => // ignore
  case _ => Behavior.interpretSignal(lastBehavior, ctx, PostStop)
```

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala (hunk 26 above)
Comment (helena) [reply akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala]: Nice.

**Review by helena — APPROVED** [review]
LGTM!
